capsules/src/mlx90614.rs

```diff
@@ -155,55 +151,60 @@ impl<'a> i2c::I2CClient for Mlx90614SMBus<'a> {
     }
 }
 
-impl<'a> LegacyDriver for Mlx90614SMBus<'a> {
-    fn command(&self, command_num: usize, _data1: usize, _data2: usize, _: AppId) -> ReturnCode {
+impl<'a> Driver for Mlx90614SMBus<'a> {
+    fn command(&self, command_num: usize, _data1: usize, _data2: usize, _: AppId) -> CommandResult {
         match command_num {
-            0 => ReturnCode::SUCCESS,
+            0 => CommandResult::success(),
             // Check is sensor is correctly connected
             1 => {
                 if self.state.get() == State::Idle {
                     self.is_present();
-                    ReturnCode::SUCCESS
+                    CommandResult::success()
                 } else {
-                    ReturnCode::EBUSY
+                    CommandResult::failure(ErrorCode::BUSY)
                 }
             }
             // Read Ambient Temperature
             2 => {
                 if self.state.get() == State::Idle {
                     self.read_ambient_temperature();
-                    ReturnCode::SUCCESS
+                    CommandResult::success()
                 } else {
-                    ReturnCode::EBUSY
+                    CommandResult::failure(ErrorCode::BUSY)
                 }
             }
             // Read Object Temperature
             3 => {
                 if self.state.get() == State::Idle {
                     self.read_object_temperature();
-                    ReturnCode::SUCCESS
+                    CommandResult::success()
                 } else {
-                    ReturnCode::EBUSY
+                    CommandResult::failure(ErrorCode::BUSY)
                 }
             }
             // default
-            _ => ReturnCode::ENOSUPPORT,
+            _ => CommandResult::failure(ErrorCode::NOSUPPORT),
         }
     }
 
     fn subscribe(
         &self,
         subscribe_num: usize,
-        callback: Option<Callback>,
+        callback: Callback,
         _app_id: AppId,
-    ) -> ReturnCode {
-        match subscribe_num {
+    ) -> Result<Callback, (Callback, ErrorCode)> {
+        let res = match subscribe_num {
             0 /* set the one shot callback */ => {
-				self.callback.insert(callback);
-				ReturnCode::SUCCESS
+                self.callback.replace(callback);
```

Comment:
I think you want a `mem::swap` here? Otherwise you return the passed callback, rather than the replaced one? You could assign to `callback` and that would work.

Reply:
I think this is right as-is, [replace returns the old contents of the Cell](https://doc.rust-lang.org/std/cell/struct.Cell.html#method.replace). If this is wrong, I did the same thing in #2382

Reply:
Sure, replace does that, but the return value is discarded (there's a semicolon there, and the next line returns `Ok(())`).

Reply:
Yours is correct -- you return the result if `replace` in an `Ok()`. Using `replace` is fine in general; the challenge here is that callback is inserted in the bottom `match`.

Reply:
Oh, yeah, I see that now. I think the whole bottom match and intermediate `res` variable can/should go away here, very much like #2382 probably.